network/newstream/stream.go

```diff
@@ -0,0 +1,172 @@
+
+package newstream
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/network"
+	"github.com/ethersphere/swarm/p2p/protocols"
+	"github.com/ethersphere/swarm/state"
+)
+
+// SlipStream implements node.Service
+var _ node.Service = (*SlipStream)(nil)
+var (
+	pollTime           = 1 * time.Second
+	createStreamsDelay = 50 * time.Millisecond //to avoid a race condition where we send a message to a server that hasnt set up yet
+)
+
+var SyncerSpec = &protocols.Spec{
+	Name:       "bzz-stream",
+	Version:    8,
+	MaxMsgSize: 10 * 1024 * 1024,
+	Messages: []interface{}{
+		StreamInfoReq{},
+		StreamInfoRes{},
+		GetRange{},
+		OfferedHashes{},
+		ChunkDelivery{},
+		WantedHashes{},
+	},
+}
+
+// SlipStream is the base type that handles all client/server operations on a node
+// it is instantiated once per stream protocol instance, that is, it should have
+// one instance per node
+type SlipStream struct {
+	mtx            sync.RWMutex
+	intervalsStore state.Store //every protocol would make use of this
+	peers          map[enode.ID]*Peer
+	kad            *network.Kademlia
+
+	providers map[string]StreamProvider
+
+	spec    *protocols.Spec   //this protocol's spec
+	balance protocols.Balance //implements protocols.Balance, for accounting
+	prices  protocols.Prices  //implements protocols.Prices, provides prices to accounting
+
+	quit chan struct{} // terminates registry goroutines
+}
+
+func NewSlipStream(intervalsStore state.Store, kad *network.Kademlia, providers ...StreamProvider) *SlipStream {
+	slipStream := &SlipStream{
+		intervalsStore: intervalsStore,
+		kad:            kad,
+		peers:          make(map[enode.ID]*Peer),
+		providers:      make(map[string]StreamProvider),
+		quit:           make(chan struct{}),
+	}
+
+	for _, p := range providers {
+		slipStream.providers[p.StreamName()] = p
+	}
+
+	slipStream.spec = SyncerSpec
+
+	return slipStream
+}
+
+func (s *SlipStream) getPeer(id enode.ID) *Peer {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	p := s.peers[id]
+	return p
+}
+
+func (s *SlipStream) addPeer(p *Peer) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	s.peers[p.ID()] = p
+}
+
+func (s *SlipStream) removePeer(p *Peer) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	if _, found := s.peers[p.ID()]; found {
+		log.Error("removing peer", "id", p.ID())
+		delete(s.peers, p.ID())
+		p.Left()
+	} else {
+		log.Warn("peer was marked for removal but not found", "peer", p.ID())
+	}
+}
+
+// Run is being dispatched when 2 nodes connect
+func (s *SlipStream) Run(p *p2p.Peer, rw p2p.MsgReadWriter) error {
+	peer := protocols.NewPeer(p, rw, s.spec)
+	bp := network.NewBzzPeer(peer)
+
+	np := network.NewPeer(bp, s.kad)
+	s.kad.On(np)
+	defer s.kad.Off(np)
+
+	sp := NewPeer(bp, s.intervalsStore, s.providers)
```

Comment:
I think if `sp` contained all the handlers, and not `peer`, we wouldn't have to have such a weird structure where `Peer` is instantiated with pointers to providers and intervalsStore. A `Peer` should contain only what is relevant to a peer.